Use store based barrier only for certain store types.

pritamdamania87 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#49694 Use store based barrier only for certain store types.**

The store based barrier introduced in
https://github.com/pytorch/pytorch/pull/49419 broke for certain store types.
This is a quick fix to resolve the issues for other store types.

Differential Revision: [D25668404](https://our.internmc.facebook.com/intern/diff/D25668404/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D25668404/)!

pritamdamania committed:
Use store based barrier only for certain store types.

The store based barrier introduced in
https://github.com/pytorch/pytorch/pull/49419 broke for certain store types.
This is a quick fix to resolve the issues for other store types.

Differential Revision: [D25668404](https://our.internmc.facebook.com/intern/diff/D25668404/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D25668404/)!

[ghstack-poisoned]

pritamdamania committed:
Update on "Use store based barrier only for certain store types."


The store based barrier introduced in
https://github.com/pytorch/pytorch/pull/49419 broke for certain store types.
This is a quick fix to resolve the issues for other store types.

Differential Revision: [D25668404](https://our.internmc.facebook.com/intern/diff/D25668404/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D25668404/)!

[ghstack-poisoned]

rohan-varma commented on torch/distributed/distributed_c10d.py:
```diff
@@ -496,7 +499,11 @@ def init_process_group(backend,
+    if backend == Backend.MPI or not (
+        isinstance(store, TCPStore) or
+        isinstance(store, FileStore) or
+        isinstance(store, HashStore)
```
nit: add these to a list `store_allowlist` and then just check `type(store) not in store_allowlist` so we don't duplicate the isinstance checks below?

rohan-varma reviewed (approved):
LGTM, although CI issue is real, looks like HashStore isn't found on windows

pritamdamania committed:
Update on "Use store based barrier only for certain store types."


The store based barrier introduced in
https://github.com/pytorch/pytorch/pull/49419 broke for certain store types.
This is a quick fix to resolve the issues for other store types.

Differential Revision: [D25668404](https://our.internmc.facebook.com/intern/diff/D25668404/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D25668404/)!

[ghstack-poisoned]
